collector/exporter.go

```diff
@@ -79,79 +76,37 @@ func New(dsn string, scrapers []Scraper) *Exporter {
 	return &Exporter{
 		dsn:      dsn,
 		scrapers: scrapers,
-		totalScrapes: prometheus.NewCounter(prometheus.CounterOpts{
-			Namespace: namespace,
-			Subsystem: exporter,
-			Name:      "scrapes_total",
-			Help:      "Total number of times MySQL was scraped for metrics.",
-		}),
-		scrapeErrors: prometheus.NewCounterVec(prometheus.CounterOpts{
-			Namespace: namespace,
-			Subsystem: exporter,
-			Name:      "scrape_errors_total",
-			Help:      "Total number of times an error occurred scraping a MySQL.",
-		}, []string{"collector"}),
-		error: prometheus.NewGauge(prometheus.GaugeOpts{
-			Namespace: namespace,
-			Subsystem: exporter,
-			Name:      "last_scrape_error",
-			Help:      "Whether the last scrape of metrics from MySQL resulted in an error (1 for error, 0 for success).",
-		}),
-		mysqldUp: prometheus.NewGauge(prometheus.GaugeOpts{
-			Namespace: namespace,
-			Name:      "up",
-			Help:      "Whether the MySQL server is up.",
-		}),
+		metrics:  metrics,
 	}
 }
 
 // Describe implements prometheus.Collector.
 func (e *Exporter) Describe(ch chan<- *prometheus.Desc) {
-	// We cannot know in advance what metrics the exporter will generate
-	// from MySQL. So we use the poor man's describe method: Run a collect
-	// and send the descriptors of all the collected metrics. The problem
-	// here is that we need to connect to the MySQL DB. If it is currently
-	// unavailable, the descriptors will be incomplete. Since this is a
-	// stand-alone exporter and not used as a library within other code
-	// implementing additional metrics, the worst that can happen is that we
-	// don't detect inconsistent metrics created by this exporter
-	// itself. Also, a change in the monitored MySQL instance may change the
-	// exported metrics during the runtime of the exporter.
-
-	metricCh := make(chan prometheus.Metric)
-	doneCh := make(chan struct{})
-
-	go func() {
-		for m := range metricCh {
-			ch <- m.Desc()
-		}
-		close(doneCh)
-	}()
-
-	e.Collect(metricCh)
-	close(metricCh)
-	<-doneCh
+	ch <- e.metrics.TotalScrapes.Desc()
+	ch <- e.metrics.Error.Desc()
+	e.metrics.ScrapeErrors.Describe(ch)
+	ch <- e.metrics.MySQLUP.Desc()
```

Comment:
So we basically don't even try to register most of the metrics descs for this collector anymore? That's probably fine as it was best-effort anyway and we only have this one collector, but was surprised it wasn't explicitly mentioned.

Reply:
I've actually made a comment about it but looks like I forgot to press "Send" ;)
Yes. I think it doesn't make sense but your expertise here guys is highly required. To be honest I don't see when registering those descriptions could be useful, maybe if you write your own app, then you register all metrics first, they get cached, so you can just focus on sending metrics. But for standalone exporter where we explicitly say we can't determine all metrics I don't see the reason to do that. Any idea what the lack of registration of descriptions could impact? I didn't looked into implementation details so again, your knowledge here could be useful ;)

Also registering those metrics made sense as long as collector was registered once, on binary start, so first http request was calling `Describe()` but next request were using cached descriptions and `Describe()` was not called. Now we register collector every scrape, so `Describe` is called every http request anyway, cached descriptions are not re-used.

Initially I thought my idea was ugly hack, but I checked later and node_exporter does the same: https://github.com/prometheus/node_exporter/blob/master/collector/collector.go#L106-L110 They also in similar way register collector every http request https://github.com/prometheus/node_exporter/blob/master/node_exporter.go#L47

Reply:
Yeah, it's fine in this case.